### PR TITLE
fix: allow params from previous tasks results

### PR DESCRIPTION
**Pull request by JordanGoasdoue**
Hello,

I first want to thank you so much for the help to allow the upgrade to 1.25 @tomhobson @msvticket and all others i'm forgetting.

I am using params with as value the results from previous tasks.
It worked in jx 3.10.83, but since I have upgraded to 3.10.86 in order to upgrade my cluster to 1.25 it broke

I know that jx uses resolver will be removed, but in order to prepare the right way, I need to keep a little more the uses inside my jx3 catalog, and so this PR would help be a lot.

I was able to reproduce with this PipelineRun that is generated after `jx uses` resolver.

```
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: jx3-test-save-show-results
spec:
  pipelineSpec:
    tasks:
      - name: save-results
        params:
        - name: value_to_save
          value: "test_demo"
        taskSpec:
          params:
          - name: value_to_save
            type: string
            default: ""
          results:
          - name: "result_value"
            description: "result_value"
          stepTemplate:
            imagePullPolicy: IfNotPresent
            resources:
              # override limits for all containers here
              limits: {}
            workingDir: /workspace/source
          steps:
            - image: uses:jx3-catalog/tasks/test-previous-results-params/0.1/save-results.yaml@4cf2604452bde421e90b3cb79f04ecb81d452188
              name: ""
      - name: display-results
        runAfter:
          - save-results
        params:
        - name: value_to_show
          value: "$(tasks.save-results.results.result_value)"
        taskSpec:
          params:
          - name: value_to_show
            type: string
            default: ""
          stepTemplate:
            imagePullPolicy: IfNotPresent
            resources:
              # override limits for all containers here
              limits: {}
            workingDir: /workspace/source
          steps:
            - image: uses:jx3-catalog/tasks/test-previous-results-params/0.1/show-results.yaml@4cf2604452bde421e90b3cb79f04ecb81d452188
              name: ""
  podTemplate: {}
  serviceAccountName: tekton-bot
  timeout: 00h30m0s
status: {}
```

with these 2 tasks that are resolved by the jx uses resolver:
```
apiVersion: tekton.dev/v1beta1
kind: Task
metadata:
  name: save-results
  labels:
    app.kubernetes.io/version: "0.1"
spec:
  params:
  - name: value_to_save
    type: string
    default: ""
  results:
  - name: "result_value"
    description: "result_value"
  stepTemplate:
    imagePullPolicy: IfNotPresent
    resources:
      # override limits for all containers here
      limits: {}
    workingDir: /workspace/source
  steps:
    - name: save-result
      image: ubuntu:jammy
      script: |
        #!/usr/bin/env sh
        set -ex
        result_value="$(params.value_to_save)"
        echo "results.result_value.path is $(results.result_value.path)"
        echo -n "${result_value}" > $(results.result_value.path)
        echo "result_value is ${result_value}"

```

```
apiVersion: tekton.dev/v1beta1
kind: Task
metadata:
  name: show-results
  labels:
    app.kubernetes.io/version: "0.1"
spec:
  params:
  - name: value_to_show
    type: string
    default: ""
  stepTemplate:
    imagePullPolicy: IfNotPresent
    resources:
      # override limits for all containers here
      limits: {}
    workingDir: /workspace/source
  steps:
    - name: show-result
      image: ubuntu:jammy
      script: |
        #!/usr/bin/env sh
        set -ex
        echo "tasks.save-results.results.result_value is $(tasks.save-results.results.result_value)"
        value_to_show="$(params.value_to_show)"
        echo "value_to_show is ${value_to_show}"

```

Before this PullRequest was done, all of this was resolved to this PipelineRun with some cleaning from myself to make it more visible:

```
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: failed
  namespace: jx
spec:
  params:
  - name: value_to_save
    value: ""
  - name: value_to_show
    value: ""
  pipelineSpec:
    params:
    - default: ""
      name: value_to_save
      type: string
    - default: ""
      name: value_to_show
      type: string
    results:
    - description: result_value
      name: result_value
      value: $(tasks.save-results.results.result_value)
    tasks:
    - name: save-results
      params:
      - name: value_to_save
        value: test_demo
      taskSpec:
        params:
        - default: ""
          name: value_to_save
          type: string
        results:
        - description: result_value
          name: result_value
          type: string
        stepTemplate:
          imagePullPolicy: IfNotPresent
          name: ""
          workingDir: /workspace/source
        steps:
        - image: ubuntu:jammy
          name: save-result
          script: |
            #!/usr/bin/env sh
            set -ex
            result_value="$(params.value_to_save)"
            echo "results.result_value.path is $(results.result_value.path)"
            echo -n "${result_value}" > $(results.result_value.path)
            echo "result_value is ${result_value}"
    - name: display-results
      params:
      - name: value_to_show
        value: $(tasks.save-results.results.result_value)
      runAfter:
      - save-results
      taskSpec:
        params:
        - default: ""
          name: value_to_show
          type: string
        stepTemplate:
          imagePullPolicy: IfNotPresent
          name: ""
          workingDir: /workspace/source
        steps:
        - image: ubuntu:jammy
          name: show-result
          script: |
            #!/usr/bin/env sh
            set -ex
            echo "tasks.save-results.results.result_value is $(tasks.save-results.results.result_value)"
            value_to_show="$(params.value_to_show)"
            echo "value_to_show is ${value_to_show}"
  serviceAccountName: tekton-bot
  timeout: 30m0s

```
Here the problem is that it seems with tekton 0.41.0+ we don't need the `PipelineRun.spec.params value_to_show` and the `PipelineRun.spec.pipelineSpec.params value_to_show` because value_to_show is just a task previous result as shown here:

``` 
params:
  - name: value_to_show
    value: $(tasks.save-results.results.result_value)
```

After this PullRequest, it is now working, this a PipelineRun that is generated because of jx uses resolver
```
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: success
  namespace: jx
spec:
  params:
  - name: value_to_save
    value: ""
  pipelineSpec:
    params:
    - default: ""
      name: value_to_save
    results:
    - description: result_value
      name: result_value
      value: $(tasks.save-results.results.result_value)
    tasks:
    - name: save-results
      params:
      - name: value_to_save
        value: test_demo
      taskSpec:
        params:
        - default: ""
          name: value_to_save
          type: string
        results:
        - description: result_value
          name: result_value
          type: string
        stepTemplate:
          imagePullPolicy: IfNotPresent
          name: ""
          workingDir: /workspace/source
        steps:
        - image: ubuntu:jammy
          name: save-result
          script: |
            #!/usr/bin/env sh
            set -ex
            result_value="$(params.value_to_save)"
            echo "results.result_value.path is $(results.result_value.path)"
            echo -n "${result_value}" > $(results.result_value.path)
            echo "result_value is ${result_value}"
    - name: display-results
      params:
      - name: value_to_show
        value: $(tasks.save-results.results.result_value)
      runAfter:
      - save-results
      taskSpec:
        params:
        - default: ""
          name: value_to_show
          type: string
        stepTemplate:
          imagePullPolicy: IfNotPresent
          name: ""
          workingDir: /workspace/source
        steps:
        - image: ubuntu:jammy
          name: show-result
          script: |
            #!/usr/bin/env sh
            set -ex
            echo "tasks.save-results.results.result_value is $(tasks.save-results.results.result_value)"
            value_to_show="$(params.value_to_show)"
            echo "value_to_show is ${value_to_show}"
  serviceAccountName: tekton-bot
  timeout: 30m0s
```

### CODE REVIEW

File: pkg/triggerconfig/inrepo/params.go [hunk 1]
```diff
@@ -22,18 +22,30 @@ type UseLocation struct {
...
+	pipelineTasksAndFinally = append(pipelineTasksAndFinally, loc.PipelineSpec.Finally...)
+	for _, pipelineTask := range pipelineTasksAndFinally {
+		params := pipelineTask.Params
+		for _, param := range params {
+			paramValStr := param.Value.StringVal
+			if strings.HasPrefix(paramValStr, "$(tasks.") && strings.Contains(paramValStr, ".results.") && strings.HasSuffix(paramValStr, ")") {
+				areParamsFromTasksResults[param.Name] = true
+			}
+		}
+	}
```
Comment (tomhobson) [inline pkg/triggerconfig/inrepo/params.go:36]: Can we reduce the nesting here?

(let's shove it into a function?)

**Review by tomhobson — CHANGES_REQUESTED** [review]

File: pkg/triggerconfig/inrepo/params.go (hunk 1 above)
Comment (JordanGoasdoue) [reply pkg/triggerconfig/inrepo/params.go:35]: done

Comment (JordanGoasdoue) [reply pkg/triggerconfig/inrepo/params.go:35]: I have also added a test :ok_hand:

**Review by tomhobson — APPROVED** [review]
It's going to be migrated away from in future, so it's only temporary. I'm sure we could split hairs but I think that wouldn't be very time efficient.

We'll have to look at the Tekton support for this @JordanGoasdoue, to make sure that the new way would work for you.

Happy to merge this for now though so that we're fully supporting your implementation. Thanks for the PR!!

**Comment by JordanGoasdoue** [comment]
Thank you @tomhobson , I totally agree with you here, and again, well done for the support, it's really appreciated

**Pull request merged**